### PR TITLE
Fix #195: Added MigrateMojo for migrating projects from FMP

### LINKED ISSUE #195: Plugin MigrateMojo to automate migration from FMP to JKube
## Description

In order to increase the adoption of the plugin we can create a `MigrateMojo` that will automatically update any `pom.xml` so that it uses Eclipse JKube instead of Fabric8 Maven Plugin.

Basically, any time the command `mvn org.eclipse.jkube:kubernetes-maven-plugin:migrate` is run in a Maven project root directory, the `pom.xml` file is parsed and appropriate substitutions are made.

Ideally, MigrateMojo should be "smart" enough to see if `fabric8-maven-plugin` should be replaced by `kubernetes-maven-plugin` or `openshift-maven-plugin`.

If the previous requirement is not possible or very difficult to implement, we can have an OpenshiftMigrateMojo that extends the basic MigrateMojo and overrides `getArtifactId` method to return `openshift-maven-plugin` instead of the super implementation `kubernetes-maven-plugin`.

The plugin should try to consider all configurations, flags and properties and rename them to match the new ones.

## See also
- There was a [MigrateMojo](https://github.com/fabric8io/fabric8/blob/874aeddd493c16133366d3ec824685d07479dbdf/fabric8-maven-plugin/src/main/java/io/fabric8/maven/MigrateMojo.java) in FMP v3, maybe there's something useful there (or maybe not)

### PR DESCRIPTION
**Pull request by rohanKanojia**
Fix #195

### CODE REVIEW

File: kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/MigrateMojo.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+public class MigrateMojo extends AbstractJKubeMojo {
+    private static final String PLUGIN_ARTIFACT_ID = "kubernetes-maven-plugin";
+    private static final String PLUGIN_GROUP_ID = "org.eclipse.jkube";
+
+    @Override
+    public void executeInternal() throws MojoExecutionException {
+        log = createLogger(null);
+        File projectBasedir = project.getBasedir();
+        try {
+            MigrateService migrateService = new MigrateService(projectBasedir, PLUGIN_GROUP_ID, getPluginArtifactId(), MavenUtil.getVersion(PLUGIN_GROUP_ID, getPluginArtifactId()), log);
```
Comment (manusa) [inline kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/MigrateMojo.java:36]: Service shouldn't be instantiated it should be provided by the JKubeServiceHub

File: jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/XMLUtilTest.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    @Test
+    public void testReadXML() throws URISyntaxException, IOException, SAXException, ParserConfigurationException {
+        // Given
+        File sampleXML = new File(getClass().getResource("/test-project/pom.xml").toURI());
+
+        // When
+        Document document = XMLUtil.readXML(sampleXML);
+
+        // Then
+        assertNotNull(document);
```
Comment (manusa) [inline jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/XMLUtilTest.java:49]: Add at least a verification to make sure that the unmarshaled document actually matches with the content of the document

File: jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/XMLUtilTest.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    @Test
+    public void testXMLWrite() throws URISyntaxException, IOException, ParserConfigurationException, SAXException, TransformerException, XPathExpressionException {
```
Comment (manusa) [inline jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/XMLUtilTest.java:53]: // GIven
This test should start by creating a Document with code only
// When
Write the document to a file (single statement)
// Then
Compare the written file to a an expected file

File: jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/XMLUtil.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        return (Node) xpath.compile(xPathExpression).evaluate(doc, XPathConstants.NODE);
+    }
+
+    public static String getNodeValueFromDocument(Document doc, String xPathExpression) throws XPathExpressionException {
+        XPathFactory xPathfactory = XPathFactory.newInstance();
+        XPath xpath = xPathfactory.newXPath();
+        return (String) xpath.compile(xPathExpression).evaluate(doc, XPathConstants.STRING);
+    }
+
+    public static int findNodeInList(NodeList nodeList, String nodeName, String desiredValue) {
```
Comment (manusa) [inline jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/XMLUtil.java:81]: You can use XPath queries to do this for you

https://www.w3schools.com/xml/xpath_syntax.asp

https://stackoverflow.com/questions/32959791/filtering-xml-document-using-xpath-in-java

**Comment by manusa** [comment]
In general, most of the filtering operations you are performing can be accomplished by XPath, try to use that as it will be more readable and probably more efficient.

File: jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/service/MigrateService.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+            if (!isRenamed) {
+                logger.warn("Unable to rename resource fragment directory in project");
+            } else {
+                logger.info("Renamed " + DEFAULT_FABRIC8_RESOURCE_FRAGMENT_DIRECTORY + " to " + DEFAULT_RESOURCE_FRAGMENT_DIRECTORY);
+            }
+        }
+    }
+
+    private void modifyFMPPropertiesInsidePom(Document dom) throws XPathExpressionException {
+        Node propertiesNode = XMLUtil.getNodeFromDocument(dom, "/project/properties");
```
Comment (dev-gaur) [inline jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/service/MigrateService.java:133]: `"/project/properties"` can also be converted into a constant string.

File: jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/XMLUtil.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        XPath xpath = xPathfactory.newXPath();
+        return (String) xpath.compile(xPathExpression).evaluate(doc, XPathConstants.STRING);
+    }
+
+    private static DocumentBuilderFactory getDocumentBuilderFactory() throws ParserConfigurationException {
+        DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
+        documentBuilderFactory.setAttribute(XMLConstants.FEATURE_SECURE_PROCESSING, true);
+        documentBuilderFactory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "");
+        documentBuilderFactory.setAttribute(XMLConstants.ACCESS_EXTERNAL_SCHEMA, "");
+        documentBuilderFactory.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false);
```
Comment (dev-gaur) [inline jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/XMLUtil.java:80]: These urls can be converted into an array of strings and then we can use a loop for setting them here..

better than hardcoding

**Pull request merged**